src/arizona_template_renderer.erl

```diff
@@ -0,0 +1,320 @@
+-module(arizona_template_renderer).
+
+%% --------------------------------------------------------------------
+%% API function exports
+%% --------------------------------------------------------------------
+
+-export([client_render/2]).
+-export([server_render/2]).
+-export([render_changes/4]).
+
+%
+
+-ignore_xref([client_render/2]).
+-ignore_xref([server_render/2]).
+-ignore_xref([render_changes/4]).
+
+%% --------------------------------------------------------------------
+%% Types (and their exports)
+%% --------------------------------------------------------------------
+
+-type assigns() :: #{atom() := term()}.
+-export_type([assigns/0]).
+
+%% --------------------------------------------------------------------
+%% API function definitions
+%% --------------------------------------------------------------------
+
+-spec client_render(Block, Assigns) -> Rendered
+    when Block :: arizona_template_compiler:block(),
+         Assigns :: assigns(),
+         Rendered :: iolist().
+client_render(Block, Assigns) ->
+    render_block(Block, Assigns).
+
+-spec server_render(Block, Assigns) -> Result
+    when Block :: arizona_template_compiler:block(),
+         Assigns :: assigns(),
+         Result :: {ok, {Rendered, Sockets}} | {error, timeout},
+         Rendered :: iolist(),
+         Sockets :: #{arizona_template_compiler:changeable_id() := arizona_socket:t()}.
+server_render(Block, Assigns) ->
+    Self = self(),
+    Pid = spawn(fun() ->
+        Rendered = server_render_block(Block, Assigns, Self),
+        Self ! {self(), {done, Rendered}}
+    end),
+    Timeout = 5_000,
+    server_render_loop(Pid, Timeout, _Sockets = []).
+
+-spec render_changes(Target, Block, ChangesVars, Assigns) -> Changes
+    when Target :: arizona_template_compiler:changeable_id(),
+         Block :: arizona_template_compiler:block(),
+         ChangesVars :: [atom()],
+         Assigns :: assigns(),
+         Changes :: [[arizona_template_compiler:changeable_id() | binary()]].
+% NOTE
+% The correct type for 'Changes' should be:
+% > [[arizona_template_compiler:changeable_id(), binary()]].
+% But Erlang does not allow a "fixed" list type.
+% Why this instead of a proplist, for example?
+% This struct will be sent to the client via JSON, and proplists are not encoded
+% out of the box. A list of lists is the low cost and more straightforward here.
```

Comment:
Comment is not required; typespec says it all. I'm thinking it's Arizona code that deals with the return value (client-side), right? So it's on us to keep compatibility...

Reply:
[af10642](https://github.com/arizona-framework/arizona/pull/168/commits/af10642884ab5abd31658290c40956650681c79d)